File: android/src/com/android/tools/idea/structure/dialog/SidePanel.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.structure.dialog;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.roots.ui.configuration.SidePanelCountLabel;
import com.intellij.openapi.roots.ui.configuration.SidePanelSeparator;
import com.intellij.openapi.ui.popup.ListItemDescriptor;
import com.intellij.ui.Gray;
import com.intellij.ui.JBColor;
import com.intellij.ui.SeparatorWithText;
import com.intellij.ui.components.JBList;
import com.intellij.ui.components.panels.NonOpaquePanel;
import com.intellij.ui.navigation.History;
import com.intellij.ui.navigation.Place;
import com.intellij.ui.navigation.Place.Navigator;
import com.intellij.ui.popup.list.GroupedItemsListRenderer;
import com.intellij.util.ui.EmptyIcon;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import java.awt.*;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static com.intellij.ui.ScrollPaneFactory.createScrollPane;
import static com.intellij.util.ui.UIUtil.SIDE_PANEL_BACKGROUND;
import static javax.swing.ListSelectionModel.SINGLE_SELECTION;

public class SidePanel extends JPanel {
  @NotNull private final JBList myList;
  @NotNull private final DefaultListModel<Place> myListModel;
  @NotNull private final Navigator myNavigator;

  @NotNull private final List<Place> myPlaces = Lists.newArrayList();
  @NotNull private final Map<Integer, String> mySeparatorByIndex = Maps.newHashMap();
  @NotNull private final Map<Place, Presentation> myPresentationByPlace = Maps.newHashMap();

  @NotNull private final History myHistory;

  public SidePanel(@NotNull Navigator navigator, @NotNull History history) {
    super(new BorderLayout());
    myHistory = history;
    myNavigator = navigator;

    myListModel = new DefaultListModel<>();
    myList = new JBList(myListModel);
    myList.setBackground(SIDE_PANEL_BACKGROUND);
    myList.setBorder(new EmptyBorder(5, 0, 0, 0));

    ListItemDescriptor<Place> descriptor = new ListItemDescriptor<Place>() {
      @Override
      public String getTextFor(Place place) {
        return myPresentationByPlace.get(place).getText();
      }

      @Override
      public String getTooltipFor(Place place) {
        return getTextFor(place);
      }

      @Override
      public Icon getIconFor(Place place) {
        return EmptyIcon.create(16, 20);
      }

      @Override
      public boolean hasSeparatorAboveOf(Place value) {
        return getSeparatorAbove(value) != null;
      }

      @Override
      public String getCaptionAboveOf(Place value) {
        return getSeparatorAbove(value);
      }
    };

    myList.setCellRenderer(new GroupedItemsListRenderer(descriptor) {
      JPanel myExtraPanel;
      SidePanelCountLabel myCountLabel;
      CellRendererPane myValidationParent = new CellRendererPane();
      {
        mySeparatorComponent.setCaptionCentered(false);
        myList.add(myValidationParent);
      }

      @Override
      protected Color getForeground() {
        return new JBColor(Gray._60, Gray._140);
      }

      @Override
      protected SeparatorWithText createSeparator() {
        return new SidePanelSeparator();
      }

      @Override
      protected void layout() {
        myRendererComponent.add(mySeparatorComponent, BorderLayout.NORTH);
        myExtraPanel.add(myComponent, BorderLayout.CENTER);
        myExtraPanel.add(myCountLabel, BorderLayout.EAST);
        myRendererComponent.add(myExtraPanel, BorderLayout.CENTER);
      }

      @Override
      public Component getListCellRendererComponent(JList list, Object value, int index, boolean isSelected, boolean cellHasFocus) {
        layout();
        myCountLabel.setText("");
        Component component = super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);
        if (value instanceof Place) {
          Place place = (Place)value;
          Object category = place.getPath("category");
          if (category instanceof CounterDisplayConfigurable) {
            CounterDisplayConfigurable configurable = (CounterDisplayConfigurable)category;
            int count = configurable.getCount();
            if (count > 0) {
              myCountLabel.setSelected(isSelected);
              myCountLabel.setText(count > 100 ? "100+" : String.valueOf(count));
            }
          }
        }
        return component;
      }

      @Override
      protected JComponent createItemComponent() {
        myExtraPanel = new NonOpaquePanel(new BorderLayout());
        myCountLabel = new SidePanelCountLabel();
        JComponent component = super.createItemComponent();

        myTextLabel.setForeground(Gray._240);
        myTextLabel.setOpaque(true);

        return component;
      }

      @Override
      protected Color getBackground() {
        return SIDE_PANEL_BACKGROUND;
      }
    });

    add(createScrollPane(myList, true), BorderLayout.CENTER);

    myList.setSelectionMode(SINGLE_SELECTION);
    myList.addListSelectionListener(e -> {
      if (e.getValueIsAdjusting()) return;
      Object value = myList.getSelectedValue();
      if (value != null) {
        myNavigator.navigateTo(((Place)value), false);
      }
    });
  }

  @NotNull
  public JBList getList() {
    return myList;
  }

  public void addPlace(@NotNull Place place, @NotNull Presentation presentation) {
    myListModel.addElement(place);
    myPlaces.add(place);
    myPresentationByPlace.put(place, presentation);
    revalidate();
    repaint();
  }

  public void clear() {
    myListModel.clear();
    myPlaces.clear();
    myPresentationByPlace.clear();
    mySeparatorByIndex.clear();
  }

  public void addSeparator(@NotNull String text) {
    mySeparatorByIndex.put(myPlaces.size(), text);
  }

  @Nullable
  public String getSeparatorAbove(@NotNull Place place) {
    return mySeparatorByIndex.get(myPlaces.indexOf(place));
  }

  @NotNull
  public Collection<Place> getPlaces() {
    return myPlaces;
  }

  public void select(@NotNull Place place) {
    myList.setSelectedValue(place, true);
  }
}

File: android/src/com/android/tools/idea/structure/dialog/SidePanel.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.structure.dialog

import com.google.common.annotations.VisibleForTesting
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.roots.ui.configuration.SidePanelSeparator
import com.intellij.openapi.ui.popup.ListItemDescriptor
import com.intellij.openapi.util.text.StringUtil
import com.intellij.ui.AbstractExpandableItemsHandler
import com.intellij.ui.ExpandableItemsHandler
import com.intellij.ui.Gray
import com.intellij.ui.JBColor
import com.intellij.ui.ScrollPaneFactory.createScrollPane
import com.intellij.ui.SeparatorWithText
import com.intellij.ui.components.JBList
import com.intellij.ui.components.panels.NonOpaquePanel
import com.intellij.ui.navigation.History
import com.intellij.ui.navigation.Place
import com.intellij.ui.navigation.Place.Navigator
import com.intellij.ui.popup.list.GroupedItemsListRenderer
import com.intellij.ui.scale.JBUIScale
import com.intellij.util.ui.EmptyIcon
import com.intellij.util.ui.GraphicsUtil
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.UIUtil
import com.intellij.util.ui.UIUtil.SIDE_PANEL_BACKGROUND
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Font
import java.awt.Graphics
import javax.swing.CellRendererPane
import javax.swing.DefaultListModel
import javax.swing.Icon
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JList
import javax.swing.JPanel
import javax.swing.ListCellRenderer
import javax.swing.ListSelectionModel.SINGLE_SELECTION
import javax.swing.border.EmptyBorder

class SidePanel(private val myNavigator: Navigator, private val myHistory: History) : JPanel(BorderLayout()) {
  data class ProblemStats(val count: Int, val containsErrors: Boolean)

  data class PlaceData(
    val place: Place,
    val separator: String?,
    val presentation: Presentation,
    val countProvider: (() -> ProblemStats)?
  )

  private val listModel: DefaultListModel<PlaceData> = DefaultListModel()

  @VisibleForTesting
  val descriptor: ListItemDescriptor<PlaceData> = object : ListItemDescriptor<PlaceData> {
    override fun getTextFor(place: PlaceData): String? = place.presentation.text
    override fun getTooltipFor(place: PlaceData): String? = null
    override fun getIconFor(place: PlaceData): Icon = JBUIScale.scaleIcon(EmptyIcon.create(16, 20))
    override fun hasSeparatorAboveOf(value: PlaceData): Boolean = value.separator != null
    override fun getCaptionAboveOf(value: PlaceData): String? = value.separator
  }

  private val cellRenderer: ListCellRenderer<PlaceData> = object : GroupedItemsListRenderer<PlaceData>(descriptor) {
    var extraPanel: JPanel? = null
    var countLabel: JLabel? = null
    private var containsErrors: Boolean = false
    private var isSelected: Boolean = false
    private val textBorders = JBUI.Borders.empty(2, 5, 2, 6 + 6)
    private val emptyBorders = JBUI.Borders.empty()

    init {
      mySeparatorComponent.setCaptionCentered(false)
    }

    override fun getForeground(): Color = JBColor(Gray._60, Gray._140)

    override fun createSeparator(): SeparatorWithText = SidePanelSeparator()

    override fun layout() {
      countLabel?.border = if (countLabel?.text.isNullOrEmpty()) emptyBorders else textBorders
      val extraPanel = extraPanel ?: return
      myRendererComponent.add(mySeparatorComponent, BorderLayout.NORTH)
      extraPanel.add(myComponent, BorderLayout.CENTER)
      extraPanel.add(countLabel, BorderLayout.EAST)
      myRendererComponent.add(this.extraPanel, BorderLayout.CENTER)
    }

    override fun getListCellRendererComponent(
      list: JList<out PlaceData>,
      value: PlaceData,
      index: Int,
      isSelected: Boolean,
      cellHasFocus: Boolean
    ): Component {
      val component = super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
      updateCountLabel(isSelected, value)
      layout()
      if (UIUtil.isClientPropertyTrue(list, ExpandableItemsHandler.EXPANDED_RENDERER)) {
        val bounds = list.getCellBounds(index, index)
        bounds.setSize(component.preferredSize.getWidth().toInt(), bounds.getHeight().toInt())
        extraPanel?.let {
          AbstractExpandableItemsHandler.setRelativeBounds(component, bounds, it, validationParent)
          it.setSize(it.preferredSize.getWidth().toInt(), it.height)
          UIUtil.putClientProperty(it, ExpandableItemsHandler.USE_RENDERER_BOUNDS, true)
          return it
        }
      }

      return component
    }

    override fun createItemComponent(): JComponent {
      extraPanel = NonOpaquePanel(BorderLayout())
      countLabel = object : JLabel() {
        init {
          font = UIUtil.getListFont().deriveFont(Font.BOLD)
        }

        override fun paintComponent(g: Graphics) {
          g.color = if (isSelected) UIUtil.getListSelectionBackground(true) else UIUtil.SIDE_PANEL_BACKGROUND
          g.fillRect(0, 0, width, height)
          if (StringUtil.isEmpty(text)) return
          val deepBlue = JBColor(Color(0x97A4B2), Color(92, 98, 113))
          g.color = when {
            isSelected -> Gray._255.withAlpha(if (UIUtil.isUnderDarcula()) 100 else 220)
            containsErrors -> JBColor.RED.darker()
            else -> deepBlue
          }
          val config = GraphicsUtil.setupAAPainting(g)
          g.fillRoundRect(0, 3, width - 6 - 1, height - 6, height - 6, height - 6)
          config.restore()
          foreground = when {
            isSelected && containsErrors -> JBColor.RED.darker()
            isSelected -> deepBlue.darker()
            else -> UIUtil.getListForeground(true, true)
          }
          super.paintComponent(g)
        }

      }
      val component = super.createItemComponent()

      myTextLabel.foreground = Gray._240
      myTextLabel.isOpaque = true

      return component
    }

    override fun getBackground(): Color = SIDE_PANEL_BACKGROUND

    private fun updateCountLabel(isSelected: Boolean, value: PlaceData) {
      val countLabel = countLabel ?: return
      val problemStats = value.countProvider?.invoke()
      val count = problemStats?.count ?: 0

      this.isSelected = isSelected
      containsErrors = problemStats?.containsErrors ?: false
      countLabel.text = when {
        count == 0 -> ""
        count > 100 -> "100+"
        else -> count.toString()
      }
    }
  }

  val validationParent = CellRendererPane()
  val list: JBList<PlaceData> = JBList(listModel)
    .also {
      it.background = SIDE_PANEL_BACKGROUND
      it.border = EmptyBorder(5, 0, 0, 0)
      it.selectionMode = SINGLE_SELECTION
      it.cellRenderer = cellRenderer
      it.add(validationParent)

      it.addListSelectionListener { e ->
        if (e.valueIsAdjusting) return@addListSelectionListener
        it.selectedValue?.let { value ->
          myNavigator.navigateTo(value.place, false)
        }
      }
    }

  private var pendingSeparator: String? = null

  init {
    add(createScrollPane(list, true), BorderLayout.CENTER)
  }

  fun addPlace(place: Place, presentation: Presentation, counterProvider: (() -> ProblemStats)?) {
    listModel.addElement(PlaceData(place, pendingSeparator, presentation, counterProvider))
    pendingSeparator = null
  }

  fun clear() {
    listModel.clear()
    pendingSeparator = null
  }

  fun addSeparator(text: String) {
    pendingSeparator = text
  }

  fun select(place: Place) {
    list.setSelectedValue(listModel.elements().asSequence().find { it.place == place }, true)
  }
}
